File: src/codegen/rollback.py
from __future__ import annotations

import shutil
import sys
from pathlib import Path
from typing import Sequence

from codegen.backup import list_timestamps
from codegen.errors import EXIT_BLOCK_FAILURE, EXIT_OK, EXIT_STARTUP


def run(
    paths: Sequence[Path],
    *,
    timestamp: str | None,
    list_only: bool,
    backup_dir: Path,
) -> int:
    if not backup_dir.exists():
        print(f"codegen rollback: backup dir not found: {backup_dir}", file=sys.stderr)
        return EXIT_STARTUP

    targets = [p.resolve() for p in paths] if paths else _all_backed_up(backup_dir)

    if list_only:
        for t in targets:
            tss = list_timestamps(t, backup_dir)
            if tss:
                print(f"{t}:")
                for ts in tss:
                    print(f"  {ts}")
        return EXIT_OK

    had_failure = False
    for t in targets:
        tss = list_timestamps(t, backup_dir)
        if not tss:
            print(f"codegen rollback: no backup for {t}", file=sys.stderr)
            had_failure = True
            continue

        ts = timestamp or tss[-1]
        if ts not in tss:
            print(f"codegen rollback: timestamp {ts!r} not found for {t}", file=sys.stderr)
            had_failure = True
            continue

        t_abs = t.resolve()
        backup_dir_abs = backup_dir.resolve()
        root = backup_dir_abs.parent

        try:
            rel = t_abs.relative_to(root)
        except ValueError:
            rel = Path(t_abs.name)

        src = backup_dir_abs / rel.parent / ts / t_abs.name
        if not src.exists():
            print(f"codegen rollback: backup file missing: {src}", file=sys.stderr)
            had_failure = True
            continue

        try:
            shutil.copy2(src, t)
        except OSError as e:
            print(f"codegen rollback: failed to restore {t}: {e}", file=sys.stderr)
            had_failure = True

    return EXIT_BLOCK_FAILURE if had_failure else EXIT_OK


def _all_backed_up(backup_dir: Path) -> list[Path]:
    files: list[Path] = []
    if not backup_dir.exists():
        return files
    for child in backup_dir.rglob("*"):
        if child.is_file():
            # Structure: backup_dir / rel_parent / timestamp / filename
            # Original path: backup_dir.parent / rel_parent / filename
            try:
                rel = child.relative_to(backup_dir)
                parts = rel.parts
                if len(parts) >= 3:
                    orig = backup_dir.parent / Path(*parts[:-2]) / parts[-1]
                    if orig not in files:
                        files.append(orig)
            except ValueError:
                pass
    return files

File: cpp/src/rollback.cpp
#include "rollback.hpp"
#include "backup.hpp"
#include "errors.hpp"

#include <algorithm>
#include <iostream>
#include <system_error>

static std::vector<fs::path> all_backed_up(const fs::path& backup_dir) {
    std::vector<fs::path> files;
    std::error_code ec;
    if (!fs::exists(backup_dir, ec)) return files;

    for (auto& entry : fs::recursive_directory_iterator(backup_dir, ec)) {
        if (!entry.is_regular_file()) continue;
        try {
            fs::path rel = fs::relative(entry.path(), backup_dir);
            auto parts = rel;
            // Structure: rel_parent / timestamp / filename
            // Need at least 3 components: <dir>/<ts>/<file>
            std::vector<fs::path> comps;
            for (auto& p : rel) comps.push_back(p);
            if (comps.size() >= 3) {
                fs::path orig = backup_dir.parent_path();
                for (size_t i = 0; i + 2 < comps.size(); i++)
                    orig /= comps[i];
                orig /= comps.back();
                if (std::find(files.begin(), files.end(), orig) == files.end())
                    files.push_back(orig);
            }
        } catch (...) {}
    }
    return files;
}

int run_rollback(const std::vector<fs::path>& paths,
                  const std::optional<std::string>& timestamp,
                  bool list_only,
                  const fs::path& backup_dir)
{
    std::error_code ec;
    if (!fs::exists(backup_dir, ec)) {
        std::cerr << "codegen rollback: backup dir not found: "
                  << backup_dir.string() << "\n";
        return EXIT_STARTUP;
    }

    std::vector<fs::path> targets;
    if (paths.empty()) {
        targets = all_backed_up(backup_dir);
    } else {
        for (auto& p : paths)
            targets.push_back(fs::absolute(p));
    }

    if (list_only) {
        for (auto& t : targets) {
            auto tss = list_timestamps(t, backup_dir);
            if (!tss.empty()) {
                std::cout << t.string() << ":\n";
                for (auto& ts : tss)
                    std::cout << "  " << ts << "\n";
            }
        }
        return EXIT_OK;
    }

    bool had_failure = false;
    for (auto& t : targets) {
        auto tss = list_timestamps(t, backup_dir);
        if (tss.empty()) {
            std::cerr << "codegen rollback: no backup for " << t.string() << "\n";
            had_failure = true;
            continue;
        }

        std::string ts = timestamp ? *timestamp : tss.back();
        if (std::find(tss.begin(), tss.end(), ts) == tss.end()) {
            std::cerr << "codegen rollback: timestamp '" << ts
                      << "' not found for " << t.string() << "\n";
            had_failure = true;
            continue;
        }

        fs::path t_abs     = fs::absolute(t);
        fs::path bdir_abs  = fs::absolute(backup_dir);
        fs::path root      = bdir_abs.parent_path();

        fs::path rel;
        try {
            rel = fs::relative(t_abs, root);
        } catch (...) {
            rel = t_abs.filename();
        }

        fs::path src = bdir_abs / rel.parent_path() / ts / t_abs.filename();
        if (!fs::exists(src, ec)) {
            std::cerr << "codegen rollback: backup file missing: " << src.string() << "\n";
            had_failure = true;
            continue;
        }

        fs::copy_file(src, t, fs::copy_options::overwrite_existing, ec);
        if (ec) {
            std::cerr << "codegen rollback: failed to restore " << t.string()
                      << ": " << ec.message() << "\n";
            had_failure = true;
        }
    }

    return had_failure ? EXIT_BLOCK_FAILURE : EXIT_OK;
}
